### ISSUE TITLE
async/await assertion failed: unbounded recursion

### ISSUE THREAD

**Issue by Ekleog** [rust-lang/rust#53249]
***(edit: *much* simpler reproducer at second post)***

(see edit history for the initial issue text)

Potential duplicate of https://github.com/rust-lang/rust/issues/49537 and https://github.com/rust-lang/rust/issues/51624

cc https://github.com/rust-lang/rust/issues/50547

**Comment by Ekleog** [comment]
Another simpler reproducer:
```rust
#![feature(async_await, await_macro, futures_api)]

use futures::prelude::*;

async fn __receive<WantFn, Fut>(want: WantFn) -> ()
where
    Fut: Future<Output = ()>,
    WantFn: Fn(&Box<Send + 'static>) -> Fut,
{
    await!(futures::future::lazy(|_| ()));
}

pub fn basic_spawn_receive() {
    async { await!(__receive(|_| async { () })) };
}
```

Cargo.toml:
```rust
cargo-features = ["edition"]

[package]
name = "erlust"
edition = "2018"
version = "0.1.0"

[dependencies.futures-preview]
git = "https://github.com/rust-lang-nursery/futures-rs"
rev = "1b6e8150e7339478e73b87ff3d616010bdb1ab0b"
```

Backtrace:
```rust
   Compiling erlust v0.1.0 (file:///home/ekleog/prog/erlust/erlust)                                                                                                                                      
thread 'main' panicked at 'assertion failed: !value.has_escaping_regions()', librustc/ty/sty.rs:756:9
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
   1: std::sys_common::backtrace::print
   2: std::panicking::default_hook::{{closure}}
   3: std::panicking::default_hook
   4: rustc::util::common::panic_hook
   5: std::panicking::rust_panic_with_hook
   6: std::panicking::begin_panic
   7: rustc::traits::util::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::predicate_for_trait_def
   8: rustc::infer::InferCtxt::in_snapshot
   9: <core::iter::FlatMap<I, U, F> as core::iter::iterator::Iterator>::next
  10: <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T, I>>::from_iter
  11: rustc::traits::select::SelectionContext::collect_predicates_for_types
  12: rustc::traits::select::SelectionContext::confirm_candidate
  13: rustc::infer::InferCtxt::probe
  14: rustc::traits::select::SelectionContext::evaluate_stack
  15: rustc::ty::context::tls::with_context
  16: rustc::traits::select::SelectionContext::evaluate_predicate_recursively
  17: rustc::infer::InferCtxt::probe
  18: rustc::traits::select::SelectionContext::evaluate_stack
  19: rustc::ty::context::tls::with_context
  20: rustc::traits::select::SelectionContext::evaluate_predicate_recursively
  21: rustc::infer::InferCtxt::probe
  22: rustc::traits::select::SelectionContext::evaluate_stack
  23: rustc::ty::context::tls::with_context
  24: rustc::traits::select::SelectionContext::evaluate_predicate_recursively
  25: rustc::infer::InferCtxt::probe
  26: rustc::traits::select::SelectionContext::evaluate_stack
  27: rustc::ty::context::tls::with_context
  28: rustc::traits::select::SelectionContext::evaluate_predicate_recursively
  29: rustc::infer::InferCtxt::probe
  30: rustc::traits::select::SelectionContext::evaluate_stack
  31: rustc::ty::context::tls::with_context
  32: rustc::traits::select::SelectionContext::evaluate_predicate_recursively
  33: rustc::infer::InferCtxt::probe
  34: rustc::traits::select::SelectionContext::evaluate_stack
  35: rustc::ty::context::tls::with_context
  36: rustc::traits::select::SelectionContext::evaluate_predicate_recursively
  37: rustc::infer::InferCtxt::probe
  38: rustc::traits::select::SelectionContext::evaluate_stack
  39: rustc::ty::context::tls::with_context
  40: rustc::traits::select::SelectionContext::evaluate_predicate_recursively
  41: rustc::infer::InferCtxt::probe
  42: rustc::traits::select::SelectionContext::evaluate_stack
  43: rustc::ty::context::tls::with_context
  44: rustc::traits::select::SelectionContext::evaluate_predicate_recursively
  45: rustc::infer::InferCtxt::probe
  46: rustc::traits::select::SelectionContext::evaluate_stack
  47: rustc::ty::context::tls::with_context
  48: rustc::traits::select::SelectionContext::evaluate_predicate_recursively
  49: rustc::traits::select::SelectionContext::evaluate_obligation_recursively
  50: rustc::ty::context::tls::with_related_context
  51: rustc::infer::InferCtxtBuilder::enter
  52: rustc_traits::evaluate_obligation::evaluate_obligation
  53: rustc::ty::query::__query_compute::evaluate_obligation
  54: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors<'tcx> for rustc::ty::query::queries::evaluate_obligation<'tcx>>::compute
  55: rustc::ty::context::tls::with_context
  56: rustc::dep_graph::graph::DepGraph::with_task_impl
  57: rustc::ty::context::tls::with_related_context
  58: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::force_query_with_job
  59: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::get_query
  60: rustc::traits::query::evaluate_obligation::<impl rustc::infer::InferCtxt<'cx, 'gcx, 'tcx>>::evaluate_obligation
  61: rustc::traits::type_known_to_meet_bound
  62: rustc::ty::context::tls::with_related_context
  63: rustc::infer::InferCtxtBuilder::enter
  64: rustc::ty::util::is_freeze_raw
  65: rustc::ty::query::__query_compute::is_freeze_raw
  66: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors<'tcx> for rustc::ty::query::queries::is_freeze_raw<'tcx>>::compute
  67: rustc::ty::context::tls::with_context
  68: rustc::dep_graph::graph::DepGraph::with_task_impl
  69: rustc::ty::context::tls::with_related_context
  70: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::force_query_with_job
  71: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::get_query
  72: rustc::ty::util::<impl rustc::ty::TyS<'tcx>>::is_freeze
  73: rustc_passes::rvalue_promotion::CheckCrateVisitor::handle_const_fn_call
  74: rustc_passes::rvalue_promotion::CheckCrateVisitor::check_expr
  75: rustc_passes::rvalue_promotion::CheckCrateVisitor::check_block
  76: rustc_passes::rvalue_promotion::CheckCrateVisitor::check_expr
  77: rustc_passes::rvalue_promotion::CheckCrateVisitor::check_nested_body
  78: rustc_passes::rvalue_promotion::rvalue_promotable_map
  79: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors<'tcx> for rustc::ty::query::queries::rvalue_promotable_map<'tcx>>::compute
  80: rustc::ty::context::tls::with_context
  81: rustc::dep_graph::graph::DepGraph::with_task_impl
  82: rustc::ty::context::tls::with_related_context
  83: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::force_query_with_job
  84: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::get_query
  85: rustc_passes::rvalue_promotion::const_is_rvalue_promotable_to_static
  86: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors<'tcx> for rustc::ty::query::queries::const_is_rvalue_promotable_to_static<'tcx>>::compute
  87: rustc::ty::context::tls::with_context
  88: rustc::dep_graph::graph::DepGraph::with_task_impl
  89: rustc::ty::context::tls::with_related_context
  90: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::force_query_with_job
  91: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::get_query
  92: rustc_passes::rvalue_promotion::check_crate
  93: rustc::util::common::time
  94: rustc::ty::context::tls::enter_context
  95: <std::thread::local::LocalKey<T>>::with
  96: rustc::ty::context::TyCtxt::create_and_enter
  97: rustc_driver::driver::compile_input
  98: rustc_driver::run_compiler_with_pool
  99: <scoped_tls::ScopedKey<T>>::set
query stack during panic:
#0 [evaluate_obligation] evaluating trait selection obligation `impl core::future::future::Future: std::marker::Freeze`
#1 [is_freeze_raw] computing whether `impl core::future::future::Future` is freeze
#2 [rvalue_promotable_map] checking which parts of `basic_spawn_receive` are promotable to static
#3 [const_is_rvalue_promotable_to_static] const checking if rvalue is promotable to static `basic_spawn_receive`
end of query stack

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.30.0-nightly (73c78734b 2018-08-05) running on x86_64-unknown-linux-gnu

note: compiler flags: -C debuginfo=2 -C incremental --crate-type lib

note: some of the compiler flags provided by cargo are hidden

error: Could not compile `erlust`.

To learn more, run the command again with --verbose.
```

The backtrace looks quite different though, so I guess a fix for this one should also be checked to fix the first example.

**Comment by Ekleog** [comment]
Latest nightly appears to have fixed the ICE, however it's still making unbounded recursion :)

**Comment by nikomatsakis** [comment]
Async-await traige: Marking as blocking pending investigation to understand what is happening.

**Comment by davidtwco** [comment]
Alright, so here's what I've worked out so far:

I was able to reproduce the ICE from the original code snippet using the compiler version and library versions from when the issue was reported (`nightly-2018-08-05` was the closest version I could find and I had to use patch to make Cargo download versions of libraries that worked). I was then able to confirm that the reproduction supplied in the next comment also triggered the ICE. I was then able to get rid of the external dependencies (mostly by copying in parts of that version of futures) and still get the ICE - [here's that test case](https://gist.github.com/davidtwco/b37746cfde36e66c0eb370c40a150e3a).

After that, I upgraded to `nightly-2018-08-14` as mentioned in the next comment and confirmed that the ICE had been fixed and the unbounded recursion was happening with that same test case. Then, I took that test case that was confirmed to have the problem and tried to run it on the latest nightly - I had to make some minor adjustments as the pin module had changed since August, but I couldn't get the recursion error to happen. I'm inclined to say that this has been fixed at some point - [here's what I ended up running](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=ce92b72a4748322792cd92a9ceae11f1).
